### PR TITLE
Support writing variant information from mutations-table process to variants application

### PR DESCRIPTION
**Pull request by jkokosar**
## Checklist

* [ ] Update CHANGELOG.rst for each commit separately:
  * Pay attention to write entries under the "Unreleased" section.
  * Mark all breaking changes as "**BACKWARD INCOMPATIBLE:**" and put them
    before non-breaking changes.
  * If a commit modifies a feature listed under "Unreleased" section,
    it might be sufficient to modify the existing CHANGELOG entry from previous
    commit(s).
* [ ] Bump the process version:
  * **MAJOR version (first number)**: Backward incompatible changes (changes
    that break the api/interface). Examples: renaming the input/output, adding
    mandatory input, removing input/output...
  * **MINOR version (middle number)**: add functionality or changes in a
    backwards-compatible manner. Examples: add output field, add non-mandatory
    input parameter, use a different tool that produces same results...
  * **PATCH version (last number)**: changes/bug fixes that do not affect
    the api/interface. Examples: typo fix, change/add warning messages...
* [ ] All inputs are used in process.
* [ ] All output fields have a value assigned to them.

### Processes ###

* Code that was once accepted should be kept like it is. It should be changed
  only when its functionality has changed. Small parts may be changed
  to fix bugs (by changing code only to make it nicer you might introduce a bug).
* Set the optimal number of cores that the tool can use (check if parallelization
  can be performed).
* Description of the tool should be self explanatory with links to the tool's
  homepage and publication.
* Write explanatory input labels and add short command line parameter names
  in square brackets (for example: Maximum length [``--maxlength``]).
* Where possible provide default values for the process input fields.
  If they are provided, then ``required = False`` is not needed.
* Add descriptive error and warning messages.
* Use ``Cmd()`` for each call in the process (call of a tool, a
  Bash command, a Python script or an R script) and handle return codes.
* Add ``self.progress()`` calls where appropriate.

### Commit and PR messages ###

* Each commit should be minimal (i.e. 1 change) and self-contained (including
  tests).
* Mark incomplete PRs with the [WIP] (Work In Progress) tag.
* Do not paste links to private repositories from the public one.

### Tests ###

* Add meaningful names of test files:
  * Genomes: ``genome_species.fa.gz``
  * Annotations: ``annotation_species.gtf.gz``
  * Adapters: ``adapters-name/type-source``.
* Specific files should be saved in a folder with the name of the tested tool.
  This folder should include separate subfolders for inputs and outputs.
* Keep test files' size small. If there is no other option, add the large
  test file to the ``files/large`` directory so it will be tracked with Git LFS.
* Pay attention to edge cases: what will happen if any data is
  missing, if the length of some list is 0, if a string is given where one
  expects an integer, if a file is empty, ...
* When testing workflows, try to verify workflow success by checking one
  of the last steps. Pick the one that is most likely to fail. Also test
  that all steps have 'OK' status.

### CODE REVIEW

File: resolwe_bio/tests/processes/test_support_processors.py [hunk 1]
```diff
@@ -9,6 +9,7 @@
 from resolwe_bio.models import Sample
 from resolwe_bio.utils.filter import filter_comment_lines, filter_rnaseqc_metrics
 from resolwe_bio.utils.test import KBBioProcessTestCase
+from resolwe_bio.variants.models import Variant, VariantCall, VariantExperiment
```
Comment (gregorjerse) [inline resolwe_bio/tests/processes/test_support_processors.py:12]: This does not belong in this commit.

File: resolwe_bio/process/runtime.py [hunk 2]
```diff
@@ -21,10 +21,10 @@ def __init__(self, data: Data):
...
 
-    def add_variants(self, variants: List[Dict[str, Any]]):
+    def add_variants(self, data_source: str, variants: List[Dict[str, Any]]):
         """Add variants to the database."""
-        return communicator.add_variants(variants)
+        return communicator.add_variants((data_source, variants))
 
     def add_variants_annotations(self, variants: List[Dict[str, Any]]):
         """Add variants annotations to the database."""
-        return communicator.add_variants_asnnotations(variants)
```
Comment (gregorjerse) [inline resolwe_bio/process/runtime.py:30]: This does not really belong into this commit. 

I suggest you merge the commit  `Initialize the listener plugin`  with this one and reword the title to something like `Fix bugs in the variants app`.

File: resolwe_bio/processes/support_processors/mutations_table.py [hunk 3]
```diff
@@ -240,6 +241,40 @@ def get_depth(variants_table, bam):
...
+
+    Variants are given as <allele1>/<allele2>, e.g. T/T or C/T
+
+    Encode these variants as:
+        - 0/1 for heterozygous variants
+        - 1/1 for homozygous variants
+    """
+    try:
+        allele_line = row.get("genotype", np.nan)
+        print("allele_line", allele_line)
```
Comment (gregorjerse) [inline resolwe_bio/processes/support_processors/mutations_table.py:255]: Is this here for debugging purposes?

File: resolwe_bio/processes/support_processors/mutations_table.py [hunk 4]
```diff
@@ -240,6 +241,40 @@ def get_depth(variants_table, bam):
...
+
+    if allele1 == allele2 == row["alternative"]:
+        return "1/1"
+    else:
+        return "0/1"
+
+
+def extract_alt_allele_depth(row):
+    """Extract the depth of the alternative allele from the AD field."""
+    ad_line = row.get("unfiltered_allele_depth")
```
Comment (gregorjerse) [inline resolwe_bio/processes/support_processors/mutations_table.py:273]: When key must be present, I suggest you use `row["..."]` instead of the `get` method.

File: resolwe_bio/processes/support_processors/mutations_table.py [hunk 5]
```diff
@@ -240,6 +241,40 @@ def get_depth(variants_table, bam):
...
+    _, alt = ad_line.split(",")
```
Comment (gregorjerse) [inline resolwe_bio/processes/support_processors/mutations_table.py:274]: You could also use `alt = ad_line.split(",")[1]`.

File: resolwe_bio/processes/support_processors/mutations_table.py [hunk 6]
```diff
@@ -498,3 +537,60 @@ def run(self, inputs, outputs):
...
+        columns_map = {
+            "species": "species",
+            "genome_assembly": "genome_assembly",
+            "CHROM": "chromosome",
+            "POS": "position",
+            "REF": "reference",
+            "ALT": "alternative",
+        }
+
+        if "QUAL" in requested_fields:
```
Comment (gregorjerse) [inline resolwe_bio/processes/support_processors/mutations_table.py:562]: Maybe something like this is more readable.

```
optional_mapping = {"DP": "depth", "FILTER": "filter"...}
for key, value in optional_mapping.items():
    if key in reequired_fields:
        columns_map[key] = value
```

File: resolwe_bio/processes/support_processors/mutations_table.py (hunk 3 above)
Comment (jkokosar) [reply resolwe_bio/processes/support_processors/mutations_table.py:255]: Yes, will remove.

File: resolwe_bio/processes/support_processors/mutations_table.py [hunk 7]
```diff
@@ -240,6 +241,39 @@ def get_depth(variants_table, bam):
...
+
+    Variants are given as <allele1>/<allele2>, e.g. T/T or C/T
+
+    Encode these variants as:
+        - 0/1 for heterozygous variants
+        - 1/1 for homozygous variants
+    """
+    try:
+        allele_line = row.get("genotype", np.nan)
+        allele_re = r"^([ACGT]+)/([ACGT]+)$"
```
Comment (AnzeLovse) [inline resolwe_bio/processes/support_processors/mutations_table.py:255]: I am aware this is an existing regex used by resdk but it fails to capture some of the variants. In particular `*/G` and similarly structured ones. We could add in the asterisk in the pattern `"^([ATGC*]+)/([ATGC*]+)$"`.

File: resolwe_bio/processes/support_processors/mutations_table.py [hunk 8]
```diff
@@ -498,3 +536,62 @@ def run(self, inputs, outputs):
...
+
+        output_table = output_table.rename(columns=columns_map)
+
+        if "genotype" in output_table.columns:
+            output_table["genotype"] = output_table.apply(
+                encode_variant_genotype, axis=1, result_type="reduce"
+            )
+
+        if "unfiltered_allele_depth" in output_table.columns:
+            output_table["unfiltered_allele_depth"] = output_table.apply(
```
Comment (AnzeLovse) [inline resolwe_bio/processes/support_processors/mutations_table.py:592]: This step could be vectorised to speed up the processing `output_table["unfiltered_allele_depth"].str.split(",").str[1].astype(int)`.

File: resolwe_bio/processes/support_processors/mutations_table.py [hunk 9]
```diff
@@ -498,3 +536,62 @@ def run(self, inputs, outputs):
         outputs.genes = genes
         outputs.species = inputs.variants.output.species
         outputs.build = inputs.variants.output.build
+
+        # Store the reported variants in the Variant Database
+        data_source = "RNA-Seq Variant Calling Pipeline"
```
Comment (AnzeLovse) [inline resolwe_bio/processes/support_processors/mutations_table.py:541]: Should this be an advanced parameter with the default value "RNA-Seq Variant Calling Pipeline" since any VCF file annotated with snpeff can be used as an input to this process?

File: resolwe_bio/tests/processes/test_support_processors.py [hunk 10]
```diff
@@ -1801,3 +1827,12 @@ def test_report_variants(self):
...
         )
+
+        variant_calls = VariantCall.objects.filter(
+            sample_id=ensembl_mutations.entity.id
+        )
+        self.assertEqual(len(variant_calls), 3)
+
+        variant_call_1 = variant_calls[0]
+        self.assertEqual(variant_call_1.genotype, "1/1")
+        self.assertEqual(variant_call_1.unfiltered_allele_depth, 2)
```
Comment (AnzeLovse) [inline resolwe_bio/tests/processes/test_support_processors.py:1838]: This could also test the `genotype_quality` field since it is `None` in the previous example.

File: resolwe_bio/processes/support_processors/mutations_table.py (hunk 7 above)
Comment (jkokosar) [reply resolwe_bio/processes/support_processors/mutations_table.py]: Fixed as suggested, thanks!

File: resolwe_bio/processes/support_processors/mutations_table.py [hunk 11]
```diff
@@ -498,3 +536,62 @@ def run(self, inputs, outputs):
...
+
+        output_table["species"] = inputs.variants.output.species
+        output_table["genome_assembly"] = (
+            "GRCh38"
+            if "GRCh38" in inputs.variants.output.build
+            else inputs.variants.output.build
+        )
+
+        sample_fields_map = {
```
[the remark this answers is not in the document]
Comment (jkokosar) [reply resolwe_bio/processes/support_processors/mutations_table.py]: Sure, its better this way!

File: resolwe_bio/processes/support_processors/mutations_table.py (hunk 9 above)
Comment (jkokosar) [reply resolwe_bio/processes/support_processors/mutations_table.py]: Added as an (advanced) input field, thanks!

**Review by AnzeLovse — APPROVED** [review]
All my comments have been addressed. @gregorjerse would you have another pass and approve if everything looks good.
